checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -0,0 +1,131 @@
+ * contains inference logic for owning annotations on final owning fields. It adds an @Owning
+ * annotation on a field if it finds a method that satisfies @MustCall obligation of the field along
+ * some path to the regular exit point.
+ */
+public class MustCallInferenceLogic {
+
+  /**
+   * The type factory for the Resource Leak Checker, which is used to access the Must Call Checker.
+   */
+  private final ResourceLeakAnnotatedTypeFactory typeFactory;
+
+  /** The @Owning annotation. */
+  protected final AnnotationMirror OWNING;
+
+  /** The control flow graph. */
+  private ControlFlowGraph cfg;
+
+  /**
+   * Creates a MustCallInferenceLogic. If the whole program inference is not null, the type
+   * factory's postAnalyze method would instantiate a new MustCallInferenceLogic using this
+   * constructor and then call {@link #inference()}.
+   *
+   * @param typeFactory the type factory
+   * @param cfg the ControlFlowGraph
+   */
+  MustCallInferenceLogic(ResourceLeakAnnotatedTypeFactory typeFactory, ControlFlowGraph cfg) {
+    this.typeFactory = typeFactory;
+    this.cfg = cfg;
+    OWNING = AnnotationBuilder.fromClass(this.typeFactory.getElementUtils(), Owning.class);
+  }
+
+  /**
+   * It checks method invocations for fields with non-empty @MustCall obligation along all paths to
+   * the regular exit point in the method body.
+   */
+  void inference() {
+    Set<Block> visited = new HashSet<>();
+    Deque<Block> worklist = new ArrayDeque<>();
+    Block entry = this.cfg.getEntryBlock();
+
+    worklist.add(entry);
+    visited.add(entry);
+
+    while (!worklist.isEmpty()) {
+      Block current = worklist.remove();
+
+      for (Node node : current.getNodes()) {
+        if (node instanceof MethodInvocationNode) {
+          checksOwningFieldForInvocation((MethodInvocationNode) node);
+        }
+      }
+
+      propagateRegPaths(current, visited, worklist);
+    }
+  }
+
+  /**
+   * If the receiver of {@code mNode} is a possible owning field and the method invocation satisfies
+   * the field's must call obligation, then adds owning annotation for that field.
+   *
+   * @param mNode the MethodInvocationNode
+   */
+  private void checksOwningFieldForInvocation(MethodInvocationNode mNode) {
+    Node receiver = mNode.getTarget().getReceiver();
+    if (receiver.getTree() == null) {
+      return;
+    }
+
+    Element receiverEl = TreeUtils.elementFromTree(receiver.getTree());
+
+    if (typeFactory.isElementPossibleOwningField(receiverEl)) {
+      Element method = TreeUtils.elementFromTree(mNode.getTree());
+      List<String> mustCallValues = typeFactory.getMustCallValue(receiverEl);
+
+      // Because we assumed that any must call annotation has at most one method, the following
+      // check is enough to decide whether the receiver is an owning field
+      if (mustCallValues.contains(method.getSimpleName().toString())) {
+        WholeProgramInference wpi = typeFactory.getWholeProgramInference();
```

Comment:
We don't go over exceptional successors, so we're always on paths to the regular exit point but what you suggested is clearer and I'll update it.

Reply:
Ah interesting. Then why do you need to check if the successor is the exceptional exit in the propagate method?

Reply:
you're right! We don't need to check that! 

While I was updating this part and testing the case you mentioned in the next comment, I found a bug in adding successors in the propagation method! `block.getSuccessors()` returns both regular and exceptional successors. So instead of using that I added back `getSuccessors(Block cur)` method.